packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -1,17 +1,326 @@
+      return Padding(
+        padding: const EdgeInsets.only(bottom: 8.0),
+        child: Container(
+          decoration: BoxDecoration(
+            border: Border.all(color: Theme.of(context).focusColor),
+          ),
+          child: controller.timelineMode == TimelineMode.frameBased
+              ? FrameBasedTimelineFlameChart(
+                  controller.frameBasedTimeline.data.selectedFrame,
+                  width: constraints.maxWidth,
+                  height: math.max(
+                    constraints.maxHeight,
+                    _frameBasedTimelineChartHeight(controller),
+                  ),
+                  selectionProvider: () =>
+                      controller.frameBasedTimeline.data.selectedEvent,
+                  onSelection: (e) => controller.selectTimelineEvent(e),
+                )
+              // TODO(kenz): implement full timeline flame chart.
+              : Container(
+                  color: Colors.black26,
+                  child: const Center(
+                    child: Text('TODO Full Timeline Flame Chart'),
+                  ),
+                ),
+        ),
+      );
+    });
+  }
+
+  double _frameBasedTimelineChartHeight(TimelineController controller) {
+    return (controller.frameBasedTimeline.data.displayDepth + 2) *
+            rowHeightWithPadding +
+        sectionSpacing;
+  }
+}
+
+// TODO(kenz): Abstract core flame chart logic for use in other flame charts.
+class FrameBasedTimelineFlameChart extends StatefulWidget {
+  FrameBasedTimelineFlameChart(
+    this.data, {
+    @required this.height,
+    @required double width,
+    @required this.selectionProvider,
+    @required this.onSelection,
+  })  : duration = data.time.duration,
+        startInset = sideInset,
+        totalStartingWidth = width;
+
+  final TimelineFrame data;
+
+  final Duration duration;
+
+  final double startInset;
+
+  final double totalStartingWidth;
+
+  final double height;
+
+  final TimelineEvent Function() selectionProvider;
+
+  final void Function(TimelineEvent event) onSelection;
+
+  double get startingContentWidth =>
+      totalStartingWidth - startInset - sideInset;
+
+  @override
+  FrameBasedTimelineFlameChartState createState() =>
+      FrameBasedTimelineFlameChartState();
+}
+
+class FrameBasedTimelineFlameChartState
+    extends State<FrameBasedTimelineFlameChart> with AutoDisposeMixin {
+  static const startingScrollPosition = 0.0;
+  ScrollController _scrollControllerX;
+  ScrollController _scrollControllerY;
+  double scrollOffsetX = startingScrollPosition;
+  double scrollOffsetY = startingScrollPosition;
+
+  List<FlameChartRow> rows;
+
+  TimelineController _controller;
+
+  int get gpuSectionStartRow => widget.data.uiEventFlow.depth;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _controller = Controllers.of(context).timeline;
+    autoDispose(_controller.onSelectedTimelineEvent.listen((_) {
```

Comment:
We need to call cancel() here.

Reply:
Gotcha. Landing this. Will fix in the next PR I'm about to send out.